src/interpreter.rs

```diff
@@ -1,222 +1,417 @@
+    fn execute(&self, ctx: &mut Context, m: &mut Match) -> bool;
 }
 
 impl Execute for Expression {
-    fn execute(&self, context: &mut Context, m: &mut Match) -> bool {
+    fn execute(&self, ctx: &mut Context, m: &mut Match) -> bool {
         match self {
             Expression::Logical(l) => match l.as_ref() {
-                LogicalExpression::And(l, r) => l.execute(context, m) && r.execute(context, m),
-                LogicalExpression::Or(l, r) => l.execute(context, m) || r.execute(context, m),
+                LogicalExpression::And(l, r) => l.execute(ctx, m) && r.execute(ctx, m),
+                LogicalExpression::Or(l, r) => l.execute(ctx, m) || r.execute(ctx, m),
             },
-            Expression::Predicate(p) => {
-                let lhs_values = match context.value_of(&p.lhs.var_name) {
-                    None => return false,
-                    Some(v) => v,
-                };
-
-                let (lower, any) = p.lhs.get_transformations();
-
-                for mut lhs_value in lhs_values
-                    .iter()
-                    .take(if any { lhs_values.len() } else { 1 })
-                {
-                    let lhs_value_transformed;
-
-                    if lower {
-                        match lhs_value {
-                            Value::String(s) => {
-                                lhs_value_transformed = Value::String(s.to_lowercase());
-                                lhs_value = &lhs_value_transformed;
-                            }
-                            _ => unreachable!(),
+            Expression::Predicate(p) => p.execute(ctx, m),
+        }
+    }
+}
+
+impl Execute for ast::Predicate {
+    fn execute(&self, ctx: &mut Context, m: &mut Match) -> bool {
+        let lhs_values = match ctx.value_of(&self.lhs.var_name) {
+            None => return false,
+            Some(v) => v,
+        };
+
+        let (lower, any) = self.lhs.get_transformations();
+
+        // only can be "all" or "any" mode.
+        // - all: all values must match (default)
+        // - any: ok if any any matched
+        for mut lhs_value in lhs_values.iter() {
+            let lhs_value_transformed;
+            if lower {
+                match lhs_value {
+                    Value::String(s) => {
+                        lhs_value_transformed = Value::String(s.to_lowercase());
+                        lhs_value = &lhs_value_transformed;
+                    }
+                    _ => unreachable!(),
+                }
+            }
+            let mut curr_match = false;
+            match self.op {
+                BinaryOperator::Equals => {
+                    if lhs_value == &self.rhs {
+                        m.matches
+                            .insert(self.lhs.var_name.clone(), self.rhs.clone());
+                        curr_match = true;
+                        if any {
+                            return true;
+                        }
```

Comment:
This check can be moved to before the above line, as if this is `true` then above line does not needs execution.